integrations/salesforce/subscription.go

```diff
@@ -0,0 +1,143 @@
+package salesforce
+
+import (
+	"context"
+	"sync"
+
+	pb "github.com/developerforce/pub-sub-api/go/proto"
+	"github.com/linkedin/goavro/v2"
+	"go.uber.org/zap"
+
+	// "go.autokitteh.dev/autokitteh/integrations/common"
+	"go.autokitteh.dev/autokitteh/sdk/sdktypes"
+)
+
+var (
+	// Key = Salesforce instance URL (to ensure one gRPC client per app).
+	// Writes happen during server startup, and when a user
+	// creates a new Salesforce connection in the UI.
+	pubSubClients = make(map[string]*pb.PubSubClient)
+
+	mu = &sync.Mutex{}
+)
+
+func (h handler) Subscribe(instanceURL, orgID, accessToken string, cid sdktypes.ConnectionID) {
+	// Ensure multiple users don't reference the same app at the same time.
+	mu.Lock()
+	defer mu.Unlock()
+
+	if _, ok := pubSubClients[instanceURL]; ok {
+		return
+	}
+
+	ctx := context.Background()
+	conn, err := initConn(accessToken, instanceURL, orgID, h.logger)
+	if err != nil {
+		h.logger.Error("failed to create gRPC connection", zap.Error(err))
+		return
+	}
+
+	client := pb.NewPubSubClient(conn)
+	pubSubClients[instanceURL] = &client
+
+	for _, eventType := range supportedEventTypes() {
+		go h.handleSalesForceSubscription(ctx, client, eventType, cid)
+	}
+}
+
+func (h handler) handleSalesForceSubscription(ctx context.Context, client pb.PubSubClient, topicName string, cid sdktypes.ConnectionID) {
+	const defaultBatchSize = int32(100)
+	// var latestReplayId []byte
+	numLeftToReceive := 0
+
+	stream, err := client.Subscribe(ctx)
+	if err != nil {
+		h.logger.Error("failed to create stream", zap.Error(err))
+		return
+	}
+
+	// Start receiving messages
+	for {
+		if numLeftToReceive <= 0 {
+			n, err := renewSubscription(stream, defaultBatchSize, topicName, h.logger)
+			if err != nil {
+				h.logger.Error("failed to renew subscription", zap.Error(err))
+				continue
+			}
+			numLeftToReceive = n
+		}
+
+		// Assumes that the stream will abort after 270 seconds of inactivity.
+		// https://developer.salesforce.com/docs/platform/pub-sub-api/references/methods/subscribe-rpc.html#subscribe-keepalive-behavior
+		msg, err := stream.Recv()
+		if err != nil {
+			h.logger.Error("error receiving message", zap.Error(err))
+			continue
+		}
+
+		// TODO(INT-314): Save the latest replay ID.
+		// latestReplayId = msg.GetLatestReplayId()
+
+		// Process the received message.
+		eventCount := len(msg.Events)
+		if eventCount > 0 {
+			numLeftToReceive -= eventCount
+		}
```

Comment:
Unnecessary `if` - the count can't be `< 0`, and `numLeftToReceive -= 0` is not a problem

```suggestion
		numLeftToReceive -= len(msg.Events)
```

Reply:
I think this check is necessary because this is where I receive empty keepalive messages. I don't want to count those as events because we will be renewing subscriptions too often.

Reply:
Done